Proposal: Platform-wide log drain

rjocoleman opened the pull request:
Currently [platform logging](http://docs.deis.io/en/latest/managing_deis/platform_logging/) is done via `journalctl -o short -f | ncat --ssl logs2.papertrailapp.com 23654`

There are a couple of disadvantages to this:
1. A unit needs to be created (and customised) by end users.
2. Container logs aren't prefixed with application name.

I propose adding an external drain to `logger` that would send to external syslog servers in addition to the current write to disk process.

Interface would be:

```
$ deisctl config logs set drain/host logs2.papertrailapp.com
$ deisctl config logs set drain/port 23654
```

Logger pulls this configuration out of etcd and sends the messages on.

related #980

deis-admin commented:
Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

mboersma commented:
Jenkins, add to whitelist.

rjocoleman commented:
I've added the docs - if everyone's happy with this as the interface I'll get started on logger.

gabrtv commented:
@rjocoleman to make it atomic and remove the slash in the key name, let's change the interface to:

```
$ deisctl config logs set drain=logs2.papertrailapp.com:23654
```

Otherwise, looks awesome. Let us know if we can help!

theaboutbox commented:
Perhaps I'm over-thinking it, but would you also want to add a protocol to the drain setting? Example:

    $ deisctl config logs set drain=syslog://logs2.papertrailapp.com:23654

I can think of a few advantages:

* Explicit: The setting doesn't make it obvious whether it's sending TCP, UDP, Syslog, HTTPS, etc. Also if a user tries something that's not implemented, it's easier to give an informative error message.
* Future-proof: Right after syslog working, I'd want to dig in and add tcp and tcp+tls
* Can use URI parsing libraries to get protocol, host, port, etc.

gabrtv commented:
@theaboutbox I see little downside to this. +1.

rjocoleman commented:
I've just pushed a first pass at this, feedback please (I'm pretty new to Go)

bacongobbler commented on logger/drain/drain.go:
```diff
@@ -0,0 +1,59 @@
+
+	s, err := client.Get(etcdPath+"/drain", true, false)
+	if err != nil {
+		return ""
```
It seems odd that we're silently passing an error here

rjocoleman replied:
How would you prefer this done?
etcd gives out an error `100: Key not found (/deis/logs/drain) [2674]`

rjocoleman replied:
@bacongobbler any more feedback on this?

bacongobbler replied:
sorry! This was a recent issue: #3007

rjocoleman replied:
np, I've rebased this.
I'm still not sure what to do about the silent error - it's probably preferable to every logged line firing a "drain not configured". I can't see a better way to pull from etcd.

bacongobbler replied:
neither can I. I think it's fine as-is unless we can think of a better solution. :)

bacongobbler commented on logger/syslogd/syslogd.go:
```diff
@@ -82,6 +84,9 @@ func (h *handler) mainLoop() {
 			break
 		}
+		if drain.GetDrain() != "" {
+			drain.SendToDrain(m.String(), drain.GetDrain())
```
With each call to `.GetDrain()`, a new etcd client is created. We should be re-using the client instead of spawning a new connection.

rjocoleman replied:
Good point, will adjust to:
```go
		d := drain.GetDrain()
		if d != "" {
			drain.SendToDrain(m.String(), d)
		}
```

bacongobbler commented:
Hmm, the Jenkins issue seems unrelated. I punted another job for you :)

bacongobbler commented on logger/drain/drain.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+func sendToSyslogDrain(m string, drain string) error {
+	conn, err := net.Dial("udp", drain)
```
we'll need to make a call to close the connection after we're done with it via `defer conn.Close()`

bacongobbler commented:
asides from deferring to close the connection, this LGTM! We'll still need to manually test, but this looks awesome :fireworks:

rjocoleman commented:
@bacongobbler I've added in the `defer conn.Close()` now, thanks!

bacongobbler commented on logger/drain/drain.go:
```diff
@@ -0,0 +1,60 @@
+
+func sendToSyslogDrain(m string, drain string) error {
+	conn, err := net.Dial("udp", drain)
+	defer conn.Close()
```
we need to call `defer` *after* we check if it's nil. Can't call `.Close()` on a `nil` type!

rjocoleman replied:
Oops! force push coming to fix.

bacongobbler commented:
apologies, there was a bug in the tests which is making this fail. I've personally seen this go green but we have a strict green policy. Would you mind rebasing, please? :heart:

bacongobbler commented:
also a squash would be great, too

rjocoleman commented:
> apologies, there was a bug in the tests which is making this fail. I've personally seen this go green but we have a strict green policy before merging PRs. Would you mind rebasing, please?

I've rebased and squashed, I hope Jenkins likes this now!

carmstrong commented:
This looks great. Code LGTM. Thanks @rjocoleman!

mboersma commented:
Ideally this would have tests and comments on the public funcs, but we don't yet have a test framework in the logger/syslog package and I will backfill the comments in #3118. Code LGTM.

carmstrong commented:
This works exactly as advertised! I set:
```console
$ deisctl config logs set drain=syslog://logs2.papertrailapp.com:23654
```
(which is the host/port for my test Papertrail account), and moments later log entries started appearing in my dashboard. `deis logs` also still showed the logs there, which is what we want.

Thanks @rjocoleman!

carmstrong commented:
@rjocoleman Could you please rebase this PR on master one last time? It failed to work for me locally until I had rebased on master, due to some fixes unrelated to this PR. Thanks!

rjocoleman commented:
@carmstrong sure, this is done now

wenzowski commented:
Really excited to see app logs prefixed with the app's name instead of just `sh`. grep ftw. love it.

carmstrong commented:
This is ready to ship once Jenkins passes it a last time.

rjocoleman commented:
@carmstrong not sure what happened here? Is this a Jenkins issue? `Error response from daemon: No such container: deis-cache-jenkins-test-integration-2180`

mboersma commented:
@rjocoleman yes, we're having some CI issues (some filesystem corruption on a test node). I think it's better now and I'll run this one again.

parndt commented:
Looks like Jenkins passed this a last time so this seems good to merge and ship, to me. :+1:

bacongobbler commented:
There's no tests on this one so it's a false positive, hence the "requires-manual-testing" label :)

mboersma commented:
Code LGTM.

carmstrong commented:
> There's no tests on this one so it's a false positive, hence the "requires-manual-testing" label :)

I manually tested myself. :shipit:
